### ISSUE TITLE
Bug Typescript Nigtly .js moduleResolve nodenext - usage of require not detected properly

### ISSUE THREAD

**Issue by frank-dspeed** [microsoft/TypeScript#48716]
# Bug Report

### 🔎 Search Terms

require not used

### 🕗 Version & Regression Information

- I was unable to test this on prior versions because not needed

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?target=99&ts=4.7.0-dev.20220408#code/JYWwDg9gTgLgBCCATArgGwKZwGZQiOAckVU0IG4AoAegCpbK5a4B1KAQzDAyjhgAt28DAA9IAZwzi+-LAGEAUgGU4ANx7jgEAHZwI2GVgBGwbUlMBzcY2YwIcCxnihIsDEjim4AUSUBZOBs4AAEkDDAoDABjIXc4FElPcGgYAApCbWQMAFooozRgGAxqPIKigDoAK3EAShtqSiidcXhIgEcUYEi4AF4EZHQMcqjI2IAlDA6ujFSXFPKQJ3ZylCg0Gqom7Ra4UsKsPvbOyPTy6hMzS2Hqwg3KUVd4AG9d-P24AF9yIA)

### 💻 Code

```ts
import module from 'module';
/**
 * Wrapper that exposes the CJS version of the bindings
 * to get imported in ESM 
 * @deprecated use import('node-cblite/cblite.js)
 */
const require = module.createRequire(import.meta.url);
const cblite = require('./binding.cjs');
export { cblite };
```

### 🙁 Actual behavior

src/native/binding.js:7:7 - error TS6133: 'require' is declared but its value is never read.

7 const require = module.createRequire(import.meta.url);

### 🙂 Expected behavior

require is used it can be fixed via renaming it to Require with CamelCase

i guess this is part of the nodenext resolve and that js does not preserve the require call and will translate it even when it does not come from the nativ require so require is some how now a reserved word in .js files which is not expected! as they are modules defined by package.json type module

this is even more evil when you target other environments that also got a own require implementation no matter which one require-js or the graaljs engine the deno engine anything running on raw v8 that implements require, 

require is a relativ common function name in many environments and does not match the NodeJS.Require type

## Solution 
When moduleResolution: nodenext + target: ESNext that also implicit sets preserve require in all current nodejs versions and environments require should stay require because it only exists for backward compat and is most time always a shim as we produce nodeNext from ESNext for ESNext

```
		"module": "ESNext",
		"target": "ESNext",
		"moduleResolution": "nodenext",
```

and yes there are 2 package.json near that one in the project dir and one even in the outer ../ 

both similar whole content
```
{ "type": "module" }
```

**Comment by andrewbranch** [comment]
Ah, the key piece of info is that this is a JS file. In JS we seem to assume any invocation of `require()` is the one injected into CJS modules, whether or not there is something shadowing it in scope or indeed even when you’re not in a CJS module. Yikes.

**Comment by andrewbranch** [comment]
It seems likely this has the same root cause / is basically the same bug as #48726

**Comment by frank-dspeed** [comment]
I guess your correct i code a lot of low level stuff so i code it in CJS as you know maybe there is no access to native code directly inside ESM. All the import magic is the work of the cjs-module-lexer that trys to create namedExports for the native code and CJS modules.

The engine it self is always CJS no matter if you even do node my.mjs node stays CJS as the C Code that creates the Environment is Sync by default for good reason.

when we do require('module') we get directly the memory pointer to the module C code and the createRequire can be used many times with diffrent path as you sayed 

It is localy scoped shadowing the global. 

but i also spotted many diffrent small issues while i am coding with real scopes

```
'use static' // first line entrypoint 

{

  // Creating Scope this is now not Global

}

```

that is the code that really is written into any .CJS file .cjs files are not handled like .js files nodejs did also restrict for example: SyntaxError: Block-scoped declarations (let, const, function, class) not yet supported outside strict mode 

thats why .CJS default to strict mode but it is documented no where. In general the module lexer will choose if ti is JS CJS or ESM

**Comment by frank-dspeed** [comment]
the playground link also reproduces the error but the playground link is not able to preserve the settings you need to adjust source javascript ESNext ESNext and then you get the error the playground is broken:

https://www.typescriptlang.org/play?target=99&ts=4.7.0-dev.20220408&filetype=js#code/JYWwDg9gTgLgBCCATArgGwKZwGZQiOAckVU0IG4AoAegCpbK5a4B1KAQzDAyjhgAt28DAA9IAZwzi+-LAGEAUgGU4ANx7jgEAHZwI2GVgBGwbUlMBzcY2YwIcCxnihIsDEjim4AUSUBZOBs4AAEkDDAoDABjIXc4FElPcGgYAApCbWQMAFooozRgGAxqPIKigDoAK3EAShtqSiidcXhIgEcUYEi4AF4EZHQMcqjI2IAlDA6ujFSXFPKQJ3ZylCg0Gqom7Ra4UsKsPvbOyPTy6hMzS2Hqwg3KUVd4AG9d-P24AF9yIA

**Comment by sandersn** [comment]
Like #48726, I think that the common case is best supported by the current assumption that `require` is always used for imports. @andrewbranch suggested that future module resolution modes could take into account the use of createRequire and treat it and `require` specially. This is basically a duplicate of #49726 so I'm going to close this as well; any future developments will be based on work from #50152.
